In [4]:
#Installing my libraries 
from sklearn import svm
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import MinMaxScaler
import warnings
from collections import Counter
warnings.filterwarnings('ignore')
import seaborn as sns
import matplotlib.pyplot as plt
from keras import Sequential
from keras.layers import Dense

In [5]:
df = pd.read_csv('bankloan.csv')

In [6]:
df.head()

Loan_ID Gender Married  Dependents     Education Self_Employed  \
0  LP001002   Male      No         0.0      Graduate            No   
1  LP001003   Male     Yes         1.0      Graduate            No   
2  LP001005   Male     Yes         0.0      Graduate           Yes   
3  LP001006   Male     Yes         0.0  Not Graduate            No   
4  LP001008   Male      No         0.0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

Loan_ID Gender Married  Dependents     Education Self_Employed  \
0    LP001002   Male      No         0.0      Graduate            No   
1    LP001003   Male     Yes         1.0      Graduate            No   
2    LP001005   Male     Yes         0.0      Graduate           Yes   
3    LP001006   Male     Yes         0.0  Not Graduate            No   
4    LP001008   Male      No         0.0      Graduate            No   
..        ...    ...     ...         ...           ...           ...   
594  LP002938   Male     Yes         0.0      Graduate           Yes   
595  LP002940   Male      No         0.0  Not Graduate            No   
596  LP002941   Male     Yes         2.0  Not Graduate           Yes   
597  LP002943   Male      No         NaN      Graduate            No   
598  LP002945   Male     Yes         0.0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
..               ...                ...         ...               ...   
594            16120                0.0       260.0             360.0   
595             3833                0.0       110.0             360.0   
596             6383             1000.0       187.0             360.0   
597             2987                0.0        88.0             360.0   
598             9963                0.0       180.0             360.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
1               1.0         Rural           N  
2               1.0         Urban           Y  
3               1.0         Urban           Y  
4               1.0         Urban           Y  
..              ...           ...         ...  
594             1.0         Urban           Y  
595             1.0         Rural           Y  
596             1.0         Rural           N  
597             0.0     Semiurban           N  
598             1.0         Rural           Y  

[599 rows x 13 columns]

In [28]:
df['LoanAmount'].isna().sum()# because i noticed there was NaN in the LoanAmount column

22

In [30]:
df.isna().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [31]:
df.shape

(614, 13)

since we have 614 rows.. then removing just a couple 50 rows should be okay.. because we want to get clean data right??

In [33]:
df = df.dropna()
df.isna().any()

Loan_ID              False
Gender               False
Married              False
Dependents           False
Education            False
Self_Employed        False
ApplicantIncome      False
CoapplicantIncome    False
LoanAmount           False
Loan_Amount_Term     False
Credit_History       False
Property_Area        False
Loan_Status          False
dtype: bool

We have successfully cleared all our NaN values
Let's check our shape now

In [34]:
df.shape

(480, 13)

We dropped a number of rows.. but that should not harm our models as much

In [35]:
df.head()

Loan_ID Gender Married  Dependents     Education Self_Employed  \
1  LP001003   Male     Yes         1.0      Graduate            No   
2  LP001005   Male     Yes         0.0      Graduate           Yes   
3  LP001006   Male     Yes         0.0  Not Graduate            No   
4  LP001008   Male      No         0.0      Graduate            No   
5  LP001011   Male     Yes         2.0      Graduate           Yes   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History Property_Area Loan_Status  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y  
5             1.0         Urban           Y

In [36]:
df = df.drop('Loan_ID', axis=1) #We dont need IDs when building our models
df['LoanAmount']=(df['LoanAmount']*1000).astype(int)
Counter(df['Loan_Status']) #We want to know if our dataset is imbalanced

Counter({'Y': 332, 'N': 148})

In [38]:
pre_y=df['Loan_Status']
pre_X=df.drop('Loan_Status', axis=1)


In [39]:
pre_X

Gender Married  Dependents     Education Self_Employed  ApplicantIncome  \
1      Male     Yes         1.0      Graduate            No             4583   
2      Male     Yes         0.0      Graduate           Yes             3000   
3      Male     Yes         0.0  Not Graduate            No             2583   
4      Male      No         0.0      Graduate            No             6000   
5      Male     Yes         2.0      Graduate           Yes             5417   
..      ...     ...         ...           ...           ...              ...   
609  Female      No         0.0      Graduate            No             2900   
610    Male     Yes         3.0      Graduate            No             4106   
611    Male     Yes         1.0      Graduate            No             8072   
612    Male     Yes         2.0      Graduate            No             7583   
613  Female      No         0.0      Graduate           Yes             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
1               1508.0      128000             360.0             1.0   
2                  0.0       66000             360.0             1.0   
3               2358.0      120000             360.0             1.0   
4                  0.0      141000             360.0             1.0   
5               4196.0      267000             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0       71000             360.0             1.0   
610                0.0       40000             180.0             1.0   
611              240.0      253000             360.0             1.0   
612                0.0      187000             360.0             1.0   
613                0.0      133000             360.0             0.0   

    Property_Area  
1           Rural  
2           Urban  
3           Urban  
4           Urban  
5           Urban  
..            ...  
609         Rural  
610         Rural  
611         Urban  
612         Urban  
613     Semiurban  

[480 rows x 11 columns]

In [42]:
dm_X = pd.get_dummies(pre_X).astype(int) 
dm_X

Dependents  ApplicantIncome  CoapplicantIncome  LoanAmount  \
1             1             4583               1508      128000   
2             0             3000                  0       66000   
3             0             2583               2358      120000   
4             0             6000                  0      141000   
5             2             5417               4196      267000   
..          ...              ...                ...         ...   
609           0             2900                  0       71000   
610           3             4106                  0       40000   
611           1             8072                240      253000   
612           2             7583                  0      187000   
613           0             4583                  0      133000   

     Loan_Amount_Term  Credit_History  Gender_Female  Gender_Male  Married_No  \
1                 360               1              0            1           0   
2                 360               1              0            1           0   
3                 360               1              0            1           0   
4                 360               1              0            1           1   
5                 360               1              0            1           0   
..                ...             ...            ...          ...         ...   
609               360               1              1            0           1   
610               180               1              0            1           0   
611               360               1              0            1           0   
612               360               1              0            1           0   
613               360               0              1            0           1   

     Married_Yes  Education_Graduate  Education_Not Graduate  \
1              1                   1                       0   
2              1                   1                       0   
3              1                   0                       1   
4              0                   1                       0   
5              1                   1                       0   
..           ...                 ...                     ...   
609            0                   1                       0   
610            1                   1                       0   
611            1                   1                       0   
612            1                   1                       0   
613            0                   1                       0   

     Self_Employed_No  Self_Employed_Yes  Property_Area_Rural  \
1                   1                  0                    1   
2                   0                  1                    0   
3                   1                  0                    0   
4                   1                  0                    0   
5                   0                  1                    0   
..                ...                ...                  ...   
609                 1                  0                    1   
610                 1                  0                    1   
611                 1                  0                    0   
612                 1                  0                    0   
613                 0                  1                    0   

     Property_Area_Semiurban  Property_Area_Urban  
1                          0                    0  
2                          0                    1  
3                          0                    1  
4                          0                    1  
5                          0                    1  
..                       ...                  ...  
609                        0                    0  
610                        0                    0  
611                        0                    1  
612                        0                    1  
613                        1                    0  

[480 rows x 17 columns]